cmd/clusterctl/client/cluster/topology.go

```diff
@@ -0,0 +1,613 @@
+				}
+			}
+		}
+		if performCreateValidation {
+			if err := validator.ValidateCreate(ctx, object); err != nil {
+				return errors.Wrapf(err, "failed validation of %s %s/%s", obj.GroupVersionKind().String(), obj.GetNamespace(), obj.GetName())
+			}
+		}
+
+		// Converted the defaulted and validated object back into unstructured.
+		// Note: This step also makes sure that modified object is updated into the
+		// original unstructured object.
+		if err := localScheme.Convert(object, obj, nil); err != nil {
+			return errors.Wrapf(err, "failed to convert %s to object", obj.GetKind())
+		}
+	}
+
+	return nil
+}
+
+func getClusterClasses(objs []*unstructured.Unstructured) []*unstructured.Unstructured {
+	res := make([]*unstructured.Unstructured, 0)
+	for _, obj := range objs {
+		if obj.GroupVersionKind() == clusterv1.GroupVersion.WithKind("ClusterClass") {
+			res = append(res, obj)
+		}
+	}
+	return res
+}
+
+func getClusters(objs []*unstructured.Unstructured) []*unstructured.Unstructured {
+	res := make([]*unstructured.Unstructured, 0)
+	for _, obj := range objs {
+		if obj.GroupVersionKind() == clusterv1.GroupVersion.WithKind("Cluster") {
+			res = append(res, obj)
+		}
+	}
+	return res
+}
+
+func getTemplates(objs []*unstructured.Unstructured) []*unstructured.Unstructured {
+	res := make([]*unstructured.Unstructured, 0)
+	for _, obj := range objs {
+		if strings.HasSuffix(obj.GetKind(), clusterv1.TemplateSuffix) {
+			res = append(res, obj)
+		}
+	}
+	return res
+}
+
+// generateCRDs creates mock CRD objects for all the provider specific objects in the input.
+// These CRD objects will eventually be added to the dry run client for UpdateReferenceAPIContract
+// to work as expected.
+func (t *topologyClient) generateCRDs(in *DryRunInput) []*apiextensionsv1.CustomResourceDefinition {
+	crds := []*apiextensionsv1.CustomResourceDefinition{}
+	crdMap := map[string]bool{}
+	var gvk schema.GroupVersionKind
+	for _, obj := range in.Objs {
+		gvk = obj.GroupVersionKind()
+		if strings.HasSuffix(gvk.Group, ".cluster.x-k8s.io") && !crdMap[gvk.String()] {
+			crd := &apiextensionsv1.CustomResourceDefinition{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+					Kind:       "CustomResourceDefinition",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("%s.%s", flect.Pluralize(strings.ToLower(gvk.Kind)), gvk.Group),
+					Labels: map[string]string{
+						clusterv1.GroupVersion.String(): clusterv1.GroupVersion.Version,
+					},
+				},
+			}
+			crds = append(crds, crd)
+			crdMap[gvk.String()] = true
+		}
+	}
+	return crds
+}
+
+func (t *topologyClient) affectedClusterClasses(in *DryRunInput, c client.Reader) ([]client.ObjectKey, error) {
+	ctx := context.TODO()
+	affectedClusterClasses := map[client.ObjectKey]bool{}
+	// Each of the ClusterClass that uses any of the Templates in the input is an affected ClusterClass.
+	for _, template := range getTemplates(in.Objs) {
```

Comment:
This seems to support multiple namespaces.
I was under the impression that we only allow to modify one Cluster and ClusterClass so this would all be in the same namespace?

Maybe we can avoid listing ClusterClasses for every template if we can assume we're only working on one namespace?

Reply:
Similar in affectedClusters

Reply:
Technically, the input can contain 2 templates from 2 different namespaces. This could affect clusterclasses from 2 different namespaces and clusters from 2 different namespaces. 
The input is only limited to having at most one clusterclass and/or at most one cluster, but we dont have any such limitations on templates. This limit  on the number of clusters and clusterclasses (the most common objects the users are tend to modify) is only in place to help with the edge cases the users can throw at us. 

The rest of the functions are written in a way so that they can work with more then one cluster and clusterclasses. This is so that we can drop this input limitation in the future.

That said, an optimization we can do here is to first calculate the the unique namespaces and then make the list calls. This could potentially reduce the number of list calls we make. How about we address this optimization in the follow up?